### PR TITLE
FO-863: State root data migration

### PR DESCRIPTION
**Pull request by kevinssgh**
1. Added function to adjust state root field for evm blocks.
2. The migration only executes once after a flag has been set after the first execution.
3. Added unit test to assert correct functionality.

### CODE REVIEW

File: src/components/contracts/modules/ethereum/src/impls.rs [hunk 1]
```diff
@@ -436,7 +438,79 @@ impl<C: Config> App<C> {
...
+            return Err(eg!("migrate_block_data: could not get latest block number"));
+        }
+        let mut i = U256::from(1);
+        while i <= current_block_num {
+            // get block data for block (i)
+            let id = Some(BlockId::Number(i));
+            let hash = HA256::new(Self::block_hash(ctx, id).unwrap_or_default());
+            let block = CurrentBlock::get(ctx.db.read().borrow(), &hash);
+            if block.is_none() {
+                return Err(eg!(
```
Comment (ws4charlie) [inline src/components/contracts/modules/ethereum/src/impls.rs:463]: First evm block in mainnet is `142_5000`. We actually can't fetch any block when i < 142_5000.

File: src/components/contracts/modules/ethereum/src/impls.rs [hunk 2]
```diff
@@ -436,7 +438,79 @@ impl<C: Config> App<C> {
...
+    }
+
+    pub fn migrate(ctx: &mut Context) -> Result<()> {
+        //Migrate transaction indices
+        Self::migrate_txn_idxs(ctx)?;
+        //Migrate state root data for blocks
+        let key = String::from(BLOCK_MIGRATE);
+        if !Migrated::contains_key(ctx.db.read().borrow(), key.borrow()) {
+            Self::migrate_block_data(ctx)?;
+            Migrated::insert(ctx.db.write().borrow_mut(), key.borrow(), &true)?;
```
Comment (ws4charlie) [inline src/components/contracts/modules/ethereum/src/impls.rs:512]: Do we need to commit_session() in here?

File: src/components/config/src/abci/mod.rs [hunk 3]
```diff
@@ -58,7 +58,7 @@ impl CheckPointConfig {
                                 evm_substate_height: 1733700,
                                 disable_evm_block_height: 1483286,
                                 enable_frc20_height: 1501000,
-                                evm_first_block_height: 0,
+                                evm_first_block_height: 1425000,
```
Comment (ws4charlie) [inline src/components/config/src/abci/mod.rs:61]: I tried fetching the first EVM block from mainnet using command:
`curl -X POST -H "Content-Type: application/json" -d '{"method":"eth_getBlockByNumber","params":["0x15BD0E",false],"id":1,"jsonrpc":"2.0"}' https://prod-mainnet.prod.findora.org:8545/` 

@kevinssgh It seems "0x15BD0E" (1424654) was the earliest block. Is your logic fault tolerant in this case?

Comment (kevinssgh) [reply src/components/config/src/abci/mod.rs:61]: Yes, it searches for the first non-zero state-root from 1425000. I checked the state root at that height and it was still zero.

**Review by ws4charlie — APPROVED** [review]
looks good

**Comment by tiannian** [comment]
@kevinssgh @ws4charlie I will revert this PR, please rebase this PR to main.

**Pull request merged**